src/crimson/osd/osd_operations/internal_client_request.h

```diff
@@ -0,0 +1,47 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include "crimson/osd/osd_operation.h"
+#include "crimson/osd/pg.h"
+
+namespace crimson::osd {
+
+class InternalClientRequest : public OperationT<InternalClientRequest> {
+public:
+  InternalClientRequest(Ref<PG> pg);
+  ~InternalClientRequest();
+
+  // imposed by `ShardService::start_operation<T>(...)`.
+  seastar::future<> start();
+
+protected:
+  virtual const hobject_t& get_target_oid() const = 0;
+  virtual PG::do_osd_ops_params_t get_do_osd_ops_params() const = 0;
+  virtual seastar::future<std::vector<OSDOp>> fabricate_osd_ops() = 0;
```

Comment:
nit, i'd suggest return a `future` only if necessary. in this case, can we just start with a plain `std::vector<OSDOp>`, and make if more complicated when necessary?

Reply:
also, i'd suggest use more plain words for the method name, "fabricate" implies that the ops is invented or constructed from subcomponents. probably to use something like `create_osd_ops()` would be better?

Reply:
Switched the return type and renamed the method.